File: Python/Reverse_Linked_List.py
#Node class
class Node:

    def __init__(self, data):
        self.data = data
        self.next = None

#Linked list class
class LinkedList:

    def __init__(self):
        self.head = None
	
    def push(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def reverse(self):
	#makes sure list is not empty or has only one value
        if self.head is None or self.head.next is None:
            return

        prev = None
        temp = self.head

        while temp:
            next_node = temp.next
            temp.next = prev
            prev = temp
            temp = next_node
        self.head = prev

    

    def print_list(self):
        temp = self.head
        List = []
        while temp:
            l1.append(temp.data)
            temp = temp.next
        return List

#Main to test reverse function
head = LinkedList()
head.push(77)
head.push(66)
head.push(55)
head.push(44)
head.push(33)
head.push(22)
head.push(11)
head.push(0)
print("Given list is: ", head.print_list())
head.reverse()
print("Reversed list is: ", head.print_list())

File: C/Reverse_linked_list.c
#include <stdio.h>
#include <stdlib.h>
/**********************************************************************
Program:Reverse_Linked_List
Author:Santhosh Pothineni
Purpose: This program creates a linked list then calls a the reverse_list function
to reverse the list before printing out the reveresed list.
***********************************************************************/

struct node;
typedef struct node Node;

struct node
{
	int data;
	Node* next;
};

void reverse_list(Node** pNode);


int main(int argc, char* argv[])
{
	Node* head = NULL;
	int i;
	Node* temp;
	for (i = 0; i < 10; i++)
	{
		temp = (Node*)malloc(sizeof(Node));
		if (temp == NULL)
		{
			printf("out of memory?\n");
			exit(1);
		}
		temp->data = i;
		temp->next = head;
		head = temp;
	}

	reverse_list(&head);

	temp = head;
	while (temp != NULL)
	{
		printf("%d\n", temp->data);
		temp = temp->next;
	}

	return 0;
}

void reverse_list(Node** pNode)
{
	if ((*pNode) == NULL || (*pNode)->next == NULL)
	{
		return;
	}
	Node* prev = NULL;
	Node* next = NULL;
	Node* current = *pNode;

	while (current != NULL)
	{
		next = current->next;
		current->next = prev;
		prev = current;
		current = next;
	}

	*pNode = prev;
	

}